### PR TITLE
GIT

### PR DESCRIPTION
**Pull request by cilim**
Task: [#299](https://app.productive.io/1-infinum/tasks/1047921)

3 chapters added to `Workflows/GIT` folder on `staging` environment:
- [Branches](https://beta.infinum.com/handbook/books/rails/workflows/git/branches)
- [Pull requests](https://beta.infinum.com/handbook/books/rails/workflows/git/pull-requests)
- [Clean changes](https://beta.infinum.com/handbook/books/rails/workflows/git/clean-changes)

### CODE REVIEW

**Comment by cilim** [comment]
@infinum/backend-team comments are welcome :)

File: Workflows/GIT/Pull requests.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+- What alternative implementations have you considered and why have you chosen this one, etc.?
+
+The description is a good place to include questions that came up during development.
+
+- Is this class/method name the best one?
+- Could approach B be more applicable in this case, etc.?
+
+This is also a good place to talk about the performance and security considerations if there are any.
+
+## Default Pull Request Reviewers
```
Comment (cukoreva) [inline Workflows/GIT/Pull requests.md:55]:
```suggestion
## Default pull request reviewers
```

File: Workflows/GIT/Clean changes.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
+After a while you'll get notifications about comments on your pull requests from your reviewers. Often times, you'll need to change something before the branch is ready.
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:1]:
```suggestion
After a while you'll get notifications about comments on your pull requests from your reviewers. Oftentimes you'll need to change something before the branch is ready.
```

File: Workflows/GIT/Clean changes.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+## Solving change requests
+Change requests should be applied in separate _fixup_ or _squash_ commits. Rebaseing the branch during an ongoing review is not appreciated unless there is a good reason for it, like pulling in some new and necessary changes from `master`, because it makes harder for the reviewers to know what the new changes are and what they already reviewed.
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:4]:
```suggestion
Change requests should be applied in separate _fixup_ or _squash_ commits. Rebasing the branch during an ongoing review is not appreciated unless there is a good reason for it, like pulling in some new and necessary changes from `master`, because it makes it harder for the reviewers to know what the new changes are and what they have already reviewed.
```

File: Workflows/GIT/Clean changes.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+These commits should be merged into `staging` as well when they are done.
+
+## Merging
+
+### Staging
+Once the PR is opened, the branch can be merged into `staging`, unless it contains considerable
+logic in the migrations, in which case the reviewers should prioritize reviewing the migrations
+first, and giving a thumbs up before merging.
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:13]:
```suggestion
and giving a thumbs up before merging.
```

File: Workflows/GIT/Clean changes.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+## Solving change requests
+Change requests should be applied in separate _fixup_ or _squash_ commits. Rebaseing the branch during an ongoing review is not appreciated unless there is a good reason for it, like pulling in some new and necessary changes from `master`, because it makes harder for the reviewers to know what the new changes are and what they already reviewed.
+
+These commits should be merged into `staging` as well when they are done.
+
+## Merging
+
+### Staging
+Once the PR is opened, the branch can be merged into `staging`, unless it contains considerable
+logic in the migrations, in which case the reviewers should prioritize reviewing the migrations
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:12]:
```suggestion
logic in the migrations, in which case the reviewers should prioritize reviewing the migrations,
```

File: Workflows/GIT/Clean changes.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+first, and giving a thumbs up before merging.
+
+We have 2 ways of doing branch merges to `staging`:
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:15]:
```suggestion
There are two ways of merging branches to `staging`:
```

File: Workflows/GIT/Clean changes.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+#### Cherry-picking
+
+```bash
+git switch staging
+git fetch origin staging && git pull --rebase
+git cherry-pick -n {BASE-OF-BRANCH}..{branch-name}
+```
+
+_Note_: BASE-OF-BRANCH is one commit prior of the first commit of the branch.
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:35]:
```suggestion
_Note_: BASE-OF-BRANCH is one commit prior to the first commit of the branch.
```

File: Workflows/GIT/Clean changes.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+```
+Merge-squash {branch-name}
+
+(pull request {pull-request-link})
+[optionally](cherry picked from {commit-sha})
+```
+
+Including the pull request link in the message results with a reference in the pull request feed, which is helpful because we'll know when the branch was deployed to `staging`.
+
+Make sure that everything's ok by running the specs, then push.
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:49]:
```suggestion
Make sure that everything is OK by running the specs, then push.
```

File: Workflows/GIT/Clean changes.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+#### Keep it clean
+After a while the `staging` branch history can become quite different than `master`, because of the different branch merges. You'll sometimes even find yourself pushing a couple of _fixup_ commits to `staging` if you're fixing something and then applying the changes to the `fix/branch`.
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:52]:
```suggestion
After a while the `staging` branch history can become quite different than that of the `master` branch, because of the different branch merges. You'll sometimes even find yourself pushing a couple of _fixup_ commits to `staging` if you're fixing something and then applying the changes to the `fix/branch`.
```

File: Workflows/GIT/Clean changes.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+We are resetting the `staging` branch to `master` after each sprint, or more frequently as we see fit:
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:54]:
```suggestion
The `staging` branch is reset to `master` after each sprint, or more frequently as we see fit:
```

File: Workflows/GIT/Clean changes.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+```bash
+git switch staging
+git reset --hard origin/master
+git push origin staging --force
+```
+
+### Master
+Once the PR has at least 1 approval, the branch was successfully deployed to `staging` and tested, and
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:63]:
```suggestion
Once the PR has at least one approval, the branch has been successfully deployed to `staging` and tested, and
```

File: Workflows/GIT/Clean changes.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+there are no failing specs, it can be merged into `master`.
+
+We are doing non fast forward merges to group commits of a single feature together in a meaningful
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:66]:
```suggestion
We perform non-fast-forward merges to group commits of a single feature together in a meaningful
```

File: Workflows/GIT/Clean changes.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+way. Before merging we're also rebasing the feature branch on the latest `master`, so that the git
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:67]:
```suggestion
way. Before merging we also rebase the feature branch onto the latest `master`, so that the git
```

File: Workflows/GIT/Clean changes.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+history is nice and clean, and that the latest code can be run on CI before actually merging into
+`master`. We're also squashing the review comments corrections here, so make sure the `--autosquash`
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:69]:
```suggestion
`master`. We also squash the review comments corrections here, so make sure the `--autosquash`
```

File: Workflows/GIT/Clean changes.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+option is on by default or add that flag to the rebase command.
+
+While on feature branch:
+```bash
+git fetch origin master
+git rebase -i origin/master
+git push --force
+```
+
+Then, check once again that everything is squashed correctly and continue on the `master` branch:
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:79]:
```suggestion
Then, check once again that everything is squashed correctly and continue on to the `master` branch:
```

File: Workflows/GIT/Clean changes.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+```bash
+git fetch origin master
+git switch master && git pull
+git merge --no-ff --no-edit {branch-name}
+```
+
+Make sure the history graph is nice and clean by entering the following command or similar and
+making sure that no lines "cross over".
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:88]:
```suggestion
make sure that no lines "cross over".
```

File: Workflows/GIT/Clean changes.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+| * 3345dbb Some other feature subject
+|/
+*   7eade95 Merge branch 'feature/xx-another-other-feature'
+|\
+| * 0a80385 Another feature subject
+|/
+*
+```
+
+Make sure that everything's ok by running the specs, then push.
```
Comment (cukoreva) [inline Workflows/GIT/Clean changes.md:131]:
```suggestion
Make sure that everything is OK by running the specs, then push.
```

File: Workflows/GIT/Branches.md [hunk 18]
```diff
@@ -0,0 +1,33 @@
+Our Git process is an upgraded version of [Github Flow](https://guides.github.com/introduction/flow/).
+
+![](https://ftp.infinum.co/stjepan_hadjic/git-flow-2.jpg)
+
+Instead of a single `master` branch, we use two branches to record the history of the project. The `master` branch stores code that's currently in production. The `staging` branch serves as an integration branch for features and fixes. It holds data that reflects what's currently on staging.
+
+**The master branch is always deployable.**
```
Comment (nikajukic) [inline Workflows/GIT/Branches.md:7]: I would change this to `should always be deployable` or `must always be deployable` to emphasize it a bit

File: Workflows/GIT/Branches.md [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+As we use [Productive](https://productive.io) to manage our tasks, we use **feature branches** for fixes and new features. Our branch naming convention is {type}/{task-number}{descriptive-task-name}. For example, a branch for adding authentication would be named `feature/123-authentication`.
+
+*Types*
+* _feature_ - new feature or an improvement to an existing feature
+* _fix_ - non-critical bugfix, improvement, paying the technical debt. Goes through code review process
+* _hotfix_ - time sensitive critical bugfix, that should be deployed to production as soon as possible. Not necessary that it goes through code review, but it should be revisited at a later stage, and properly fixed or improved
```
Comment (nikajukic) [inline Workflows/GIT/Branches.md:14]: Branches_-_Rails_Handbook

you need to format this a bit :D

**Comment by cilim** [comment]
Thanks for the comments @cukoreva & @nikajukic. Fixup commit pushed.

File: Workflows/GIT/Branches.md [hunk 20]
```diff
@@ -20,14 +21,14 @@ Feature branches are **always** branched out of `master`. They are also merged f
 ## Why you need the `staging` branch and why you should never merge it into `master`
 We use the `staging` server so that our QA team could test our applications in an environment that is as close to production as possible. We can also work on multiple features in parallel. Those fixes and features need to be verified by our QA team and sometimes by the client as well. Sometimes, a feature may be ready for production while others aren't and are still being worked on. In that case, the `staging` branch contains multiple features, but only one needs to end up on `master`. That is why we do not branch out of `staging` and do not merge `staging` into `master`.
 
-## Note on workflow during early development:
-While the application is still not deployed on a production server, you can omit the `staging` branch. Once the production server has been set up, and the first deploy has been up, create a `staging` branch.
+## Note on the workflow during early development:
+While the application has still not been deployed to a production server, you can omit the `staging` branch. Once the production server has been set up, and the first deploy is up, create a `staging` branch.
```
Comment (cukoreva) [inline Workflows/GIT/Branches.md:25]:
```suggestion
While the application has not been deployed to a production server yet, you can omit the `staging` branch. Once the production server has been set up, and the first deploy is up, create a `staging` branch.
```
Sorry, I've just realized this sounded a bit awkward with "still".

File: Workflows/GIT/Branches.md [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+
+As we use [Productive](https://productive.io) to manage our tasks, we use **feature branches** for fixes and new features. Our branch naming convention is {type}/{task-number}{descriptive-task-name}. For example, a branch for adding authentication would be named `feature/123-authentication`.
+
+*Types*
+
+* _feature_ - new feature or an improvement to an existing feature.
+* _fix_ - non-critical bugfix, improvement, paying the technical debt. Goes through the code review process.
+* _hotfix_ - time sensitive critical bugfix, that should be deployed to production as soon as possible. It is not necessary for it to go through a code review, but it should be revisited at a later stage, and properly fixed or improved.
+
+Feature branches are **always** branched out of `master`. They are also merged first into `staging`, and then into `master` if they are ready for production.
```
Comment (JureCindro) [inline Workflows/GIT/Branches.md:17]: I'd say that they SHOULD always be branched out of master, but not always. Sometimes you need something from a feature branch and you don't want to wait for it to be merged into Master, maybe because of something that is delaying that branch, so it would make sense to open a PR with a temporary base set to the feature branch you based on, to provide a clean difffor the PR. 
Note that this flow requres extra attentiom when merging into master as you have to change the base branch if the PR and be cautious when rebasing the feature branch in case the base branche’s history changed.

I would not include all of this in the handbook, I’d just change the wording to SHOULD so that it is not so strict 😊

File: Workflows/GIT/Branches.md [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+
+## Note on the workflow during early development:
+While the application has not been deployed to a production server yet, you can omit the `staging` branch. Once the production server has been set up, and the first deploy is up, create a `staging` branch.
+
+## Other important notes on using Git:
+**Commit messages are important**, especially since Git tracks your changes and then displays them as commits once they've been pushed to the server. By writing clear commit messages, you can make it easier for other people to follow and provide feedback.
+Commits should have a descriptive subject as well as a quick explanation of the reason for the change in the commit body. This makes it easier to check changes in the code editor as you do not have to find the pull request and open it on GitHub.
+Read more about writing proper commit messages [here](http://chris.beams.io/posts/git-commit/).
+
+**Make small commits, following the Single Responsibility Principle in Git.** This makes commits easier to review when making pull requests, and it's easier to notice what's going on when something goes wrong.
```
Comment (JureCindro) [inline Workflows/GIT/Branches.md:32]: What if we reword this to just "follow the single resposibility principle in git commits”? Small commits are more common and they can be nueacanse just as well as the larger ones.

**Comment by cilim** [comment]
@JureCindro I also do cherry-picking one by one commit. When there's too many, it's tedious so I use the range approach, but usually we keep PRs small...was also thinking whether we should add it

**Comment by melcha** [comment]
I'm thinking we should also add a link to a blog post explaining git rebase & squashing in more details, for beginners. @JureCindro @cilim thoughts?

File: Workflows/GIT/Clean changes.md [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+```bash
+git fetch origin master
+git rebase -i origin/master
+git push --force
+```
+
+Then, check once again that everything is squashed correctly and continue on to the `master` branch:
+
+```bash
+git fetch origin master
```
[the remark this answers is not in the document]
Comment (JureCindro) [reply Workflows/GIT/Clean changes.md:85]: If you do it via github there is no way of checking if the branching graph is correct. What if 2 people are mergeing into master at the same time? I'm guessing the `up-to-date` merge option has to be on?

**Comment by JureCindro** [comment]
@melcha I think this is something everybody should at least skim over https://git-scm.com/book/en/v2

&
https://git-scm.com/book/en/v2/Git-Branching-Rebasing

File: Workflows/GIT/Branches.md [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+
+Instead of a single `master` branch, we use two branches to record the history of the project. The `master` branch stores code that's currently in production. The `staging` branch serves as an integration branch for features and fixes. It holds data that reflects what's currently on staging.
+
+**The master branch must always be deployable.**
+
+As we use [Productive](https://productive.io) to manage our tasks, we use **feature branches** for fixes and new features. Our branch naming convention is {type}/{task-number}{descriptive-task-name}. For example, a branch for adding authentication would be named `feature/123-authentication`.
+
+*Types*
+
+* _feature_ - new feature or an improvement to an existing feature.
```
[the remark this answers is not in the document]
Comment (cilim) [reply Workflows/GIT/Branches.md:13]: makes sense, since the other two types have at least some reference to code reviews :)

File: Workflows/GIT/Clean changes.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (melcha) [reply Workflows/GIT/Clean changes.md:85]: Would checking the graph on the feature branch after the rebase & squash not be enough? 
We could enforce `up-to-date` option for all repos (via the project setup excel). This should be enabled regardless of the merging technique.

**Comment by melcha** [comment]
@cilim could you also include those 2 links Jure posted?

**Comment by cilim** [comment]
@melcha I've added [the links](https://github.com/infinum/rails-handbook/pull/119/files#diff-261465126551d9487b92a63906fcbbbeR135) where I find them most suitable :)

File: Workflows/GIT/Pull requests.md [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+
+Another purpose of a good description is to provide the documentation for a future reference.
+A new developer might join the project in the future. The description will help him understand
+the motivation behind the implementation of a specific feature. Please note that any important information
+from the description should also be added to the main documentation in the project as a central reference point
+where the latest state of the project is documented.
+
+A new Rails project created with the [Infinum rails template](https://github.com/infinum/default_rails_template)
+includes a [pull request description template](https://help.github.com/en/articles/creating-a-pull-request-template-for-your-repository).
+The template is defined in the `rails_project/.github/PULL_REQUEST_TEMPLATE.md` file.
```
Comment (nikajukic) [inline Workflows/GIT/Pull requests.md:17]: We talked on slack about adding a default title template as well. 

I usually use:
`Task #{task numbers} - Short description` if PR is for one or two tasks

or

`Task #tag - Short description` - if PR is for a bigger feature where all tasks are tagged with #tag

Pull_Requests_·_infinum_web-loop-community-backend

It's easier to find a PR you're looking for like this, rather than trying to match task title with PR title.

Comment (melcha) [reply Workflows/GIT/Pull requests.md:17]: Could you update the PR template in default_rails_template with this?

Comment (cilim) [reply Workflows/GIT/Pull requests.md:17]: I usually go to the Productive task and try finding the PR link there. I've always been opening PRs like that 😄 

Going to the PR list on Github and searching sounds slow if there's a lot of them.

Comment (cilim) [reply Workflows/GIT/Pull requests.md:17]: @melcha I think you can only edit the PR description, and the template is for that, not the title. I might be mistaken and there is some solution to have this automatic 🤔

File: Workflows/GIT/Clean changes.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (d4be4st) [reply Workflows/GIT/Clean changes.md:85]: > What if 2 people are mergeing into master at the same time? 

You will get an error if merge conflict occurs this way and the squash commit will fail

File: Workflows/GIT/Pull requests.md (hunk 25 above)
Comment (d4be4st) [reply Workflows/GIT/Pull requests.md:17]: I actually first go to Github and see the PRs there. Am using productive less and less these days :D

**Comment by d4be4st** [comment]
should we add some Git CLI or GUI tools here for helping with rebases, merges, etc?

Example of that would be https://github.com/extrawurst/gitui

File: Workflows/GIT/Clean changes.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (JureCindro) [reply Workflows/GIT/Clean changes.md:85]: But not if there are no merge conflicts, right? Unless there is an up-to-date option set 😄

**Comment by melcha** [comment]
> should we add some Git CLI or GUI tools here for helping with rebases, merges, etc?
> 
> Example of that would be https://github.com/extrawurst/gitui

Does anybody even use this? I wouldn't add it if nobody's even tried it. I think it's better to learn these techniques via console, and when one masters it, then they can use some gui tool if they want.
